### PR TITLE
[DLSP21] Update 12-3.md

### PR DESCRIPTION
**Pull request by makeecat**

### CODE REVIEW

File: docs/en/week12/12-3.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    - Training a policy
+- PPUU
+    - Stochastic environment
+    - Uncertainty minimisation
+    - Latent decoupling
+
+## State transition equations -- Evolution of the state
+
+Here we discuss a state transition equation where $x$ represents the state, $u$ represents control. We can formulate the state transition function in a continuous-time system where $x(t)$ is a function of continuous variable $t$.
+$\dot{x} = f(x,u)$
```
Comment (Atcold) [inline docs/en/week12/12-3.md:25]: Vectors use `\vect{}`, which calls `\bm{}`.
Also, please use `$$` and add an empty line before.
Finally, have both equations in a `\begin{align}`/`\end{align}` environment.

File: docs/en/week12/12-3.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+$\frac{\partial x(t)}{\partial t} = f(x(t),u(t))$
+<center>
```
Comment (Atcold) [inline docs/en/week12/12-3.md:27]:
```suggestion

<center>
```

File: docs/en/week12/12-3.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+In this example $\boldsymbol{x}=(x,y,\theta,s)$ is the state, $\boldsymbol{u}=(\phi,\alpha)$ is the control.
+
+$$
+\left\{\begin{array}{l}
+\dot{x}=s \cos \theta \\
+\dot{y}=s \sin \theta \\
+\dot{\theta}=\frac{s}{L} \tan \phi \\
+\dot{s}=a
+\end{array}\right.e
```
Comment (Atcold) [inline docs/en/week12/12-3.md:42]: There's an extra `e`.

File: docs/en/week12/12-3.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+Here we discuss a state transition equation where $x$ represents the state, $u$ represents control. We can formulate the state transition function in a continuous-time system where $x(t)$ is a function of continuous variable $t$.
+$\dot{x} = f(x,u)$
+$\frac{\partial x(t)}{\partial t} = f(x(t),u(t))$
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure1.png" style="background-color:#DCDCDC;" /><br>
+Figure 1: State and control illustration of a three-cycle
+</center>
+
+We use a tri-cycle as the example to study it. The orange wheel is the control $u$, $(x_c,y_c)$ is the instantaneous center of rotation. You can also have two wheels in the front. For simplicity, we use one wheel as the example.
```
Comment (Atcold) [inline docs/en/week12/12-3.md:32]: `u` is a vector.

File: docs/en/week12/12-3.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+
+We can reformulate the differential equation from continuous-time system to discreate-time system
+
+$$
+\boldsymbol{x}[t]=\boldsymbol{x}[t-1]+f(\boldsymbol{x}[t-1], \boldsymbol{u}[t]) \mathrm{d} t
+$$
+
+To be clear, we show the units of $\boldsymbol{x}, \boldsymbol{u}$.
+$$
```
Comment (Atcold) [inline docs/en/week12/12-3.md:53]:
```suggestion

$$
```

File: docs/en/week12/12-3.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+Figure 2: State Formulation
+</center>
+
+Example 1: Uniform Linear Motion: No acceleration, no steering
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure3.svg" style="background-color:#DCDCDC;" /><br>
+Figure 3: Control of Uniform Linear Motion
+</center>
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure4.svg" style="background-color:#DCDCDC;" /><br>
```
Comment (Atcold) [inline docs/en/week12/12-3.md:73]: The grid is a bit short. Please, extend to the end of the figure.
Here and below.

File: docs/en/week12/12-3.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+</center>
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure8.svg" style="background-color:#DCDCDC;" /><br>
+Figure 8: State of Sine Wave
+</center>
+
+## Kelley-Bryson algorithm
+What if we want the tri-cycle to reach a specified destination with a specified speed? 
+- This can be achieved by inference using **Kelley-Bryson algorithm**, which utilizes **backprop through time** and **gradient descent**.
+### Recap of RNN
```
Comment (Atcold) [inline docs/en/week12/12-3.md:102]:
```suggestion


### Recap of RNN
```

File: docs/en/week12/12-3.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+Example 3: Sine wave: Positive steering for the first part, negative steering for the second part
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure7.svg" style="background-color:#DCDCDC;" /><br>
+Figure 7: Control of Sine Wave
+</center>
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure8.svg" style="background-color:#DCDCDC;" /><br>
+Figure 8: State of Sine Wave
+</center>
+
```
Comment (Atcold) [inline docs/en/week12/12-3.md:98]:
```suggestion


```

File: docs/en/week12/12-3.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+- This can be achieved by inference using **Kelley-Bryson algorithm**, which utilizes **backprop through time** and **gradient descent**.
+### Recap of RNN
+We can compare the inference process here  with the training process of RNN.
+
+Below is an RNN schematic chart. We feed variable $x[t]$ and the previous state $h[t-1]$ into the predictor, while $h[0]$ is set to be zero. The predictor outputs the hidden representation $h[t]$.
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure9.png" style="background-color:#DCDCDC;" /><br>
+Figure 9: RNN schematic chart
+</center>
+### Optimal control (inference)
```
Comment (Atcold) [inline docs/en/week12/12-3.md:110]:
```suggestion


### Optimal control (inference)
```

File: docs/en/week12/12-3.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+Figure 8: State of Sine Wave
+</center>
+
+## Kelley-Bryson algorithm
+What if we want the tri-cycle to reach a specified destination with a specified speed? 
+- This can be achieved by inference using **Kelley-Bryson algorithm**, which utilizes **backprop through time** and **gradient descent**.
+### Recap of RNN
+We can compare the inference process here  with the training process of RNN.
+
+Below is an RNN schematic chart. We feed variable $x[t]$ and the previous state $h[t-1]$ into the predictor, while $h[0]$ is set to be zero. The predictor outputs the hidden representation $h[t]$.
```
Comment (Atcold) [inline docs/en/week12/12-3.md:105]: `x`, `h`, `u`, and `z` are all vectors.
Please, use `\vect{}`.

Comment (Atcold) [inline docs/en/week12/12-3.md:105]:
```suggestion
Below is an RNN schematic chart. We feed variable $x[t]$ and the previous state $h[t-1]$ into the predictor, while $h[0]$ is set to be zero vector. The predictor outputs the hidden representation $h[t]$.
```

File: docs/en/week12/12-3.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure16.png" style="background-color:#DCDCDC;" /><br>
+Figure 16: Cost Example: Softmin
+</center>
+
+Different forms of cost functions can be explored through experimentation. 
+## Optimization_Path_Planner-Notebook
+In this notebook, we use tri-cycle as an example as well. 
+
+### Define kinematic model of a tricycle $\dot{x}=f(x,u)$.
```
Comment (Atcold) [inline docs/en/week12/12-3.md:167]: `x` and `u` are vectors.

File: docs/en/week12/12-3.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+## Optimization_Path_Planner-Notebook
+In this notebook, we use tri-cycle as an example as well. 
+
+### Define kinematic model of a tricycle $\dot{x}=f(x,u)$. 
+* $x$ represents state: ($x$, $y$, $θ$, $s$)
+* $u$ represents control: ($ϕ$, $a$)
+* We feed $x[t-1]$ and $u[t]$ to obtain the next state $x[t]$
+
+
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:173]:
```suggestion
```python
```

File: docs/en/week12/12-3.md [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+### Define several cost functions
+As mentioned above, cost functions can take various forms. In this notebook, we list 5 kinds as follows: 
+* **vanilla_cost**: Focuses on the final position
+* **cost_with_target_s**: Focuses on the final position and final zero speed. 
+* **cost_sum_distances**: Focuses on the position of every step, and minimizes the mean of the distances. 
+* **cost_sum_square_distances**: Focuses on the position of every step, and minimizes the mean of squared distances.
+* **cost_logsumexp**: The distance of the closest position should be minimized. 
+
+
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:195]:
```suggestion
```python
```

File: docs/en/week12/12-3.md [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+def cost_logsumexp(state, target):
+    x_x, x_y = target
+    dists = ((state[:, 0] - x_x).pow(2) + (state[:, 1] - x_y).pow(2))#.pow(0.5)
+    return -1 * torch.logsumexp(-1 * dists, dim=0)
+```
+### Define path planning with cost
+* The optimizer is set to be SGD.
+* Time interval T is set to be 1s.
+* We need to compute every state from the initial state by the following code: 
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:223]:
```suggestion

```python
```

File: docs/en/week12/12-3.md [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+x = [torch.tensor((0, 0, 0, s),dtype=torch.float32)]
+        for t in range(1, T+1):
+            x.append(x[-1] + f(x[-1], u[t-1]) * dt)
+        x_t = torch.stack(x)
+```
+* Then compute the cost:
+  `cost = cost_f(x_t, (x_x, x_y))
+        costs.append(cost.item())`
```
Comment (Atcold) [inline docs/en/week12/12-3.md:231]: This looks like this. I'm not sure what you were aiming for.
image

File: docs/en/week12/12-3.md [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+* Implement backprop and update $u$
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:233]:
```suggestion
```python
```

**Review by Atcold — CHANGES_REQUESTED** [review]
A few minor edits. Let me know if anything in unclear.

File: docs/en/week12/12-3.md (hunk 6 above)
Comment (makeecat) [reply docs/en/week12/12-3.md:75]: Do you mean by horizontally?

File: docs/en/week12/12-3.md (hunk 1 above)
Comment (makeecat) [reply docs/en/week12/12-3.md:25]: I don't think we can use align environment, I checked the repo from last year, there was no example about align environment as well.

File: docs/en/week12/12-3.md (hunk 15 above)
Comment (Daniel-Yao-Chengdu) [reply docs/en/week12/12-3.md:239]: After we choose a specific cost (vanilla, average distance...), we can compute the cost, which is the distance of our target (x_x, x_y) and our positions x_t. The term "costs.append..." can be deleted.

File: docs/en/week12/12-3.md (hunk 1 above)
Comment (Atcold) [reply docs/en/week12/12-3.md:25]: https://katex.org/docs/supported.html#environments

File: docs/en/week12/12-3.md (hunk 6 above)
Comment (Atcold) [reply docs/en/week12/12-3.md:75]: Both ways, see below.
image

File: docs/en/week12/12-3.md (hunk 15 above)
Comment (Atcold) [reply docs/en/week12/12-3.md:239]: I meant I don't know how you wanted it to appear.
I see a in-line code split in two lines.
I didn't know whether you wanted to use a code-block or something else.

File: docs/en/week12/12-3.md (hunk 1 above)
Comment (Atcold) [reply docs/en/week12/12-3.md:25]: Also, `x` and `u` are still not bolded here.

File: docs/en/week12/12-3.md (hunk 3 above)
Comment (Atcold) [reply docs/en/week12/12-3.md:43]: Missed this?

File: docs/en/week12/12-3.md (hunk 11 above)
Comment (Atcold) [reply docs/en/week12/12-3.md:174]: They need to be shown in bold, here and everywhere else.

File: docs/en/week12/12-3.md [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+    x_x, x_y = target
+    dists = ((state[:, 0] - x_x).pow(2) + (state[:, 1] - x_y).pow(2))
+    return dists.mean()
+
+def cost_logsumexp(state, target):
+    x_x, x_y = target
+    dists = ((state[:, 0] - x_x).pow(2) + (state[:, 1] - x_y).pow(2))#.pow(0.5)
+    return -1 * torch.logsumexp(-1 * dists, dim=0)
+```
+### Define path planning with cost
```
Comment (Atcold) [inline docs/en/week12/12-3.md:226]: Two empty lines before new headings, here and elsewhere.
```suggestion


### Define path planning with cost
```

**Review by Atcold — CHANGES_REQUESTED** [review]
I think you missed a few reviews.

File: docs/en/week12/12-3.md (hunk 1 above)
Comment (makeecat) [reply docs/en/week12/12-3.md:25]: Oh I found out the wan. align does not work, aligned works. I will submit a PR addressing all issues.

File: docs/en/week12/12-3.md (hunk 15 above)
Comment (Daniel-Yao-Chengdu) [reply docs/en/week12/12-3.md:239]: I wanted to use a code-block here. I am not sure how it appears like this. I'll figure it out.

**Comment by makeecat** [comment]
I should have addressed all issues.

File: docs/en/week12/12-3.md (hunk 1 above)
Comment (Atcold) [reply docs/en/week12/12-3.md]: A PR for what? I haven't merged this yet. I'm confused.

File: docs/en/week12/12-3.md (hunk 15 above)
Comment (Atcold) [reply docs/en/week12/12-3.md]: Just use three backticks rather than a single one.

**Comment by makeecat** [comment]
> You said something about the `aligned` environment not working / you would have submitted another PR. I'm confused. Can you explain?

Sorry, it was my fault. I mean by fixing this issue by submitting the ‘fix’ commit. There will be no new PR.

**Review by Atcold — APPROVED** [review]
Great, everything looks good now!
Merging now!

**Comment by Atcold** [comment]
There are some issues with the public version we are currently investigating.
On the left there is the public version, on the right the locally run one.
image

**Pull request merged**